src/BlockTypes/LegacyTemplate.php

```diff
@@ -180,4 +190,47 @@ protected function render_archive_product() {
 		wp_reset_postdata();
 		return ob_get_clean();
 	}
+
+	/**
+	 * Get HTML markup with the right classes by attributes.
+	 * This function appends the classname at the first element that have the class attribute.
+	 * Based on the experience, all the wrapper elements have a class attribute.
+	 *
+	 * @param string $content Block content.
+	 * @param array  $block Parsed block data.
+	 * @return string Rendered block type output.
+	 */
+	public function add_alignment_class_to_wrapper( string $content, array $block ) {
+		if ( ( 'woocommerce/' . $this->block_name ) !== $block['blockName'] ) {
+			return $content;
+		}
+
+		$attributes            = (array) $block['attrs'];
+		$align_class_and_style = StyleAttributesUtils::get_align_class_and_style( $attributes );
+
+		if ( ! isset( $align_class_and_style['class'] ) ) {
+			return $content;
+		}
+
+		// Find the first tag.
+		$first_tag = '<[^<>]+>';
+		$matches   = array();
+		preg_match( $first_tag, $content, $matches );
+
+		// If there is a tag, but it doesn't have a class attribute, add the class attribute.
+		if ( isset( $matches[0] ) && strpos( $matches[0], 'class' ) === false ) {
+			$pattern_before_tag_closing = '/.+?(?=>)/';
+			$matches                    = array();
+			preg_match( $pattern_before_tag_closing, $content, $matches );
+			return preg_replace( $pattern_before_tag_closing, $matches[0] . ' class="' . $align_class_and_style['class'] . '"', $content, 1 );
```

Comment:
Can we use backreferences in [`preg_replace`](https://www.php.net/manual/en/function.preg-replace.php)? If doing so, we can remove the `preg_match` above.

Reply:
I don't think with the current regex.
Because the value of `match[0]` is a tag element without `<` and `>`, for example:
```
div data-block-name="woocommerce/legacy-template" data-template="taxonomy-product_cat" id="primary" 
```

But the regex `/.+?(?=>)/` match the text before a tag closing `>`.

Potentially we can use `str_replace`, in this way:

```			
return str_replace( $matches[0], $matches[0] . ' class="' . $align_class_and_style['class'] . '"', $content );
```
instead to use regex.

I don't have a strong opinion, what do you think?

Reply:
> Because the value of `match[0]` is a tag element without `<` and `>`, for example:

@gigitux I think we can use the fourth argument of `preg_replace` to only replace one time, you can check my example using your regex below:

```php
<?php
$re = '/.+?(?=>)/m';
$str = '<div data-block-name="woocommerce/legacy-template" data-template="taxonomy-product_cat" id="primary" ><div class="inner">';
$subst = '$0 class="alignwide"';

$result = preg_replace($re, $subst, $str, 1);

echo "The result of the substitution is <pre>" . htmlspecialchars( $result ) . "</pre>";
```

Reply:
Thanks for sharing this!